File: clase6/linded_list/single_LL/python/linked_list.py
from node import Node

class LinkedList:
    def __init__(self):
        self.head = None

    def add_first(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
        else:
            new_node.next = self.head
            self.head = new_node

    def add_last(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node

    def add_after(self, prev_node, value):
        if prev_node is None:
            return # Previous node cannot be None
        new_node = Node(value)
        new_node.next = prev_node.next
        prev_node.next = new_node
       
    def remove(self, value):
        if self.head is None:
            return False

        # If the head node is the one to be removed
        if self.head.data == value:
            self.head = self.head.next
            return True

        current = self.head
        while current.next:
            if current.next.data == value:
                current.next = current.next.next
                return True
            current = current.next
        return False  # Node not found

    def get(self, value):
        current = self.head
        while current:
            if current.data == value:
                return current
            current = current.next
        return None  # Node not found
    
    def clear(self):
        self.head = None
        self.tail = None

    def contains(self, data):
        return self.get(data) is not None
    
    def size(self):
        count = 0
        current = self.head
        while current:
            count += 1
            current = current.next
        return count
    
    def __str__(self):
        nodes = []
        current = self.head
        while current:
            nodes.append(str(current))
            current = current.next
        return " -> ".join(nodes) if nodes else "Empty List"

if __name__ == "__main__":
    ll = LinkedList()
    ll.add_last(10)
    ll.add_first(5)
    ll.add_last(15)
    print(ll)  # Output: [5] -> [10] -> [15]
    ll.add_after(ll.get(10), 12)
    print(ll)  # Output: [5] -> [10] -> [12] -> [15]
    ll.remove(5)
    print(ll)  # Output: [10] -> [12] -> [15]
    print("Size:", ll.size())  # Output: Size: 3
    print("Contains 12:", ll.contains(12))  # Output: Contains 12: True
    ll.clear()
    print(ll)  # Output: Empty List
    

File: clase6/linded_list/single_LL/java/LinkedList.java
public class LinkedList {
   /* ---- Attributes ---- */
   private Node head;

   /* ---- Constructor ---- */
   public LinkedList() {
      head = null;
   }

   /* ---- Methods ---- */

   public void addAfter(Node prevNode, int value) {
      if (prevNode == null) {
         return; // previous node cannot be null
      }
      Node newNode = new Node(value);
      newNode.next = prevNode.next;
      prevNode.next = newNode;
   }


   // Adds the given value to the front of the list.
   public void addFirst(int value) {
      Node newNode = new Node(value);
      if (head == null) {
         head = newNode;
      } else {
         newNode.next = head;
         head = newNode;
      }
   }

   // Adds the given value to the end of the list.
   public void addLast(int value) {
      Node newNode = new Node(value);
      if (head == null) {
         head = newNode;
      } else {
         Node current = head;
         while (current.next != null) {
            current = current.next;
         }
         current.next = newNode;
      }
   }

   // Returns the reference in list at given value.
   public Node get(int value) {
      Node current = head;
      while (current != null) {
         if (current.data == value) {
            return current;
         }
         current = current.next;
      }
      return null;
   }

   // Removes the first occurrence of the given value from the list.
   public boolean remove(int value) {
        if (head == null) {
            return false; // List is empty
        }
        if (head.data == value) {
            head = head.next; // Remove head
            return true;
        }
        Node current = head;
        Node prev = null;
        while (current != null) {
            if (current.data == value) {
                if (prev != null) {
                    prev.next = current.next; // Bypass the node to be removed
                }
                return true;
            }
            prev = current;
            current = current.next;
        }
        return false; // Value not found
    }

   public void clear() {
      head = null;
   }

   // Returns true if the list contains the given value, or false otherwise.
   public boolean contains(int value) {
      Node current = head;
      while (current != null) {
         if (current.data == value) {
            return true;
         }
         current = current.next;
      }
      return false;
   }

   // Returns the number of elements in the list.
   public int size() {
      int count = 0;
      Node current = head;
      while (current != null) {
         count++;
         current = current.next;
      }
      return count;
   }

   public boolean isEmpty() {
      return head == null;
   }

   @Override
   public String toString() {
      StringBuilder result = new StringBuilder();
      Node current = head;
      while (current != null) {
         result.append(current).append(" -> ");
         current = current.next;
      }
      result.append("null");
      return result.toString();
   }
}